drivers/scale_host.go

```diff
@@ -0,0 +1,403 @@
+package drivers
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	v1client "github.com/rancher/go-rancher/client"
+	"github.com/rancher/go-rancher/v2"
+	rConfig "github.com/rancher/webhook-service/config"
+	"github.com/rancher/webhook-service/model"
+)
+
+type ScaleHostDriver struct {
+}
+
+func (s *ScaleHostDriver) ValidatePayload(conf interface{}, apiClient *client.RancherClient) (int, error) {
+	config, ok := conf.(model.ScaleHost)
+	if !ok {
+		return http.StatusInternalServerError, fmt.Errorf("Can't process config")
+	}
+
+	if config.Action == "" {
+		return http.StatusBadRequest, fmt.Errorf("Scale action not provided")
+	}
+
+	if config.Action != "up" && config.Action != "down" {
+		return http.StatusBadRequest, fmt.Errorf("Invalid action %v", config.Action)
+	}
+
+	if config.Amount <= 0 {
+		return http.StatusBadRequest, fmt.Errorf("Invalid amount: %v", config.Amount)
+	}
+
+	if config.HostSelector == nil {
+		return http.StatusBadRequest, fmt.Errorf("HostSelector not provided")
+	}
+
+	if config.Min <= 0 {
+		return http.StatusBadRequest, fmt.Errorf("Minimum scale not provided/invalid")
+	}
+
+	if config.Max <= 0 {
+		return http.StatusBadRequest, fmt.Errorf("Maximum scale not provided/invalid")
+	}
+
+	if config.Min >= config.Max {
+		return http.StatusBadRequest, fmt.Errorf("Max must be greater than min")
+	}
+
+	if config.DeleteOption != "mostRecent" && config.DeleteOption != "leastRecent" {
+		return http.StatusBadRequest, fmt.Errorf("Invalid delete option %v", config.DeleteOption)
+	}
+
+	return http.StatusOK, nil
+}
+
+func (s *ScaleHostDriver) Execute(conf interface{}, apiClient *client.RancherClient, reqBody interface{}) (int, error) {
+	var currNameSuffix, baseHostName, currCloneName, suffix, key, value string
+	var count, index, newHostScale, baseHostIndex int64
+	httpClient := &http.Client{
+		Timeout: time.Second * 10,
+	}
+
+	config := &model.ScaleHost{}
+	err := mapstructure.Decode(conf, config)
+	if err != nil {
+		return http.StatusInternalServerError, errors.Wrap(err, "Couldn't unmarshal config")
+	}
+
+	action := config.Action
+	amount := config.Amount
+	min := config.Min
+	max := config.Max
+	deleteOption := config.DeleteOption
+
+	hostSelector := make(map[string]string)
+	for key, value = range config.HostSelector {
+		hostSelector[key] = value
+	}
+
+	cattleConfig := rConfig.GetConfig()
+	cattleURL := cattleConfig.CattleURL
+	u, err := url.Parse(cattleURL)
+	if err != nil {
+		panic(err)
+	}
+	cattleURL = strings.Split(cattleURL, u.Path)[0] + "/v2-beta"
```

Comment:
Per our discussion, use raw http calls as little as possible. Only use raw http at the last moment when getting the host to clone.